storage: rebalancing tests fail when file system is above 95% full

```
--- FAIL: TestStoreRangeRebalance (5.17s)
	client_raft_test.go:1312: Failed to rebalance replica within 5 seconds
```

because of the code in `selectRandom`:

```go
		if desc.Capacity.FractionUsed() > maxFractionUsedThreshold {
			continue
		}
```

These options ought to be disabled for a test like this, presumably through a refactor that attaches them to the allocator and populates them from the `StoreContext`, so that they may be influenced in tests without resorting to global hacks.

--- a/storage/engine/in_mem.go
+++ b/storage/engine/in_mem.go
@@ -28,7 +28,9 @@ type InMem struct {
 
 // NewInMem allocates and returns a new, opened InMem engine.
 func NewInMem(attrs roachpb.Attributes, cacheSize int64, stopper *stop.Stopper) InMem {
-	cache := NewRocksDBCache(cacheSize)
+	cache := NewRocksDBCache(cacheSize) // released when db closes
+	// The cache starts out with a refcount of one, and creating the engine
+	// from it adds another refcount, at which point we release one of them.
 	defer cache.Release()
 	db := InMem{
 		RocksDB: newMemRocksDB(attrs, cache, 512<<20 /* 512 MB */, stopper),

--- a/storage/engine/in_mem.go
+++ b/storage/engine/in_mem.go
@@ -28,7 +28,7 @@ type InMem struct {
 
 // NewInMem allocates and returns a new, opened InMem engine.
 func NewInMem(attrs roachpb.Attributes, cacheSize int64, stopper *stop.Stopper) InMem {
-	cache := NewRocksDBCache(cacheSize) // released when db closes
+	cache := NewRocksDBCache(cacheSize)
 	// The cache starts out with a refcount of one, and creating the engine
 	// from it adds another refcount, at which point we release one of them.
 	defer cache.Release()

--- a/storage/engine/rocksdb.go
+++ b/storage/engine/rocksdb.go
@@ -233,7 +233,9 @@ type RocksDBCache struct {
 	cache *C.DBCache
 }
 
-// NewRocksDBCache creates a new cache of the specified size.
+// NewRocksDBCache creates a new cache of the specified size. Note that the
+// cache is refcounted internally and starts out with a refcount of one (i.e.
+// Release() should be called after having used the cache).
 func NewRocksDBCache(cacheSize int64) RocksDBCache {
 	return RocksDBCache{cache: C.DBNewCache(C.uint64_t(cacheSize))}
 }
@@ -246,7 +248,8 @@ func (c RocksDBCache) ref() RocksDBCache {
 }
 
 // Release releases the cache. Note that the cache will continue to be used
-// until all of the RocksDB engines it was attached to have been closed.
+// until all of the RocksDB engines it was attached to have been closed, and
+// that RocksDB engines which use it auto-release when they close.
 func (c RocksDBCache) Release() {
 	if c.cache != nil {
 		C.DBReleaseCache(c.cache)

--- a/storage/engine/rocksdb.go
+++ b/storage/engine/rocksdb.go
@@ -306,6 +306,7 @@ func newMemRocksDB(
 		// dir: empty dir == "mem" RocksDB instance.
 		cache:          cache.ref(),
 		memtableBudget: memtableBudget,
+		maxSize:        memtableBudget,
 		stopper:        stopper,
 		deallocated:    make(chan struct{}),
 	}
@@ -474,7 +475,14 @@ func (r *RocksDB) Capacity() (roachpb.StoreCapacity, error) {
 	fileSystemUsage := gosigar.FileSystemUsage{}
 	dir := r.dir
 	if dir == "" {
-		dir = "/tmp"
+		// This is an in-memory instance. Pretend we're empty since we
+		// don't know better and only use this for testing. Using any
+		// part of the actual file system here can throw off allocator
+		// rebalancing in a hard-to-trace manner. See #7050.
+		return roachpb.StoreCapacity{
+			Capacity:  r.maxSize,
+			Available: r.maxSize,
+		}, nil
 	}
 	if err := fileSystemUsage.Get(dir); err != nil {
 		return roachpb.StoreCapacity{}, err

